federation/pkg/federation-controller/util/federated_informer.go

```diff
@@ -146,9 +164,16 @@ func NewFederatedInformer(federationClient federation_release_1_4.Interface, tar
 					return
 				}
 				if isClusterReady(oldCluster) != isClusterReady(curCluster) || !reflect.DeepEqual(oldCluster.Spec, curCluster.Spec) {
+
+					if clusterLifecycle.ClusterUnavailable != nil {
+						clusterLifecycle.ClusterUnavailable(oldCluster, federatedInformer, federatedInformer.GetTargetStore())
+					}
 					federatedInformer.deleteCluster(oldCluster)
 					if isClusterReady(curCluster) {
 						federatedInformer.addCluster(curCluster)
+						if clusterLifecycle.ClusterAvailable != nil {
+							clusterLifecycle.ClusterAvailable(curCluster, federatedInformer, federatedInformer.GetTargetStore())
+						}
```

Comment:
I'm not worried about efficiency here. I'm mostly worried about scenarios like:
"the cluster is unavailable, to let's reschedule everything from it, but in fact the cluster wasn't unavailable - it was only our library which had a shortcut for handling it".